lib/attr_encrypted/adapters/active_record.rb

```diff
@@ -43,32 +50,47 @@ def attributes_with_attr_encrypted=(*args)
           end
         end
 
+        # Saves the attr class wrappers, so AR can infer the correct type later
+        def encrypted_reflections
+          @encrypted_reflections ||= {}
+        end
+
         protected
 
           # <tt>attr_encrypted</tt> method
           def attr_encrypted(*attrs)
             super
             options = attrs.extract_options!
-            attr = attrs.pop
-            options.merge! encrypted_attributes[attr]
 
-            define_method("#{attr}_changed?") do
-              if send("#{options[:attribute]}_changed?")
-                send(attr) != send("#{attr}_was")
+            attrs.each do |attribute|
+              if (klass = options[:class])
+                self.encrypted_reflections[attribute.to_sym] = EncryptedReflection.new(klass)
               end
-            end
 
-            define_method("#{attr}_was") do
-              attr_was_options = { operation: :decrypting }
-              attr_was_options[:iv]= send("#{options[:attribute]}_iv_was") if respond_to?("#{options[:attribute]}_iv_was")
-              attr_was_options[:salt]= send("#{options[:attribute]}_salt_was") if respond_to?("#{options[:attribute]}_salt_was")
-              encrypted_attributes[attr].merge!(attr_was_options)
-              evaluated_options = evaluated_attr_encrypted_options_for(attr)
-              [:iv, :salt, :operation].each { |key| encrypted_attributes[attr].delete(key) }
-              self.class.decrypt(attr, send("#{options[:attribute]}_was"), evaluated_options)
+              attribute_options = options.merge(encrypted_attributes[attribute])
+
+              define_method("#{attribute}_changed?") do
+                public_send(attribute) != public_send("#{attribute}_was")
+              end
+
+              define_method("#{attribute}_was") do
+                attr_was_options = { operation: :decrypting }
+                attr_was_options[:iv]= public_send("#{attribute_options[:attribute]}_iv_was") if respond_to?("#{attribute_options[:attribute]}_iv_was")
+                attr_was_options[:salt]= public_send("#{attribute_options[:attribute]}_salt_was") if respond_to?("#{attribute_options[:attribute]}_salt_was")
+                encrypted_attributes[attribute].merge!(attr_was_options)
+                evaluated_options = evaluated_attr_encrypted_options_for(attribute)
+                [:iv, :salt, :operation].each { |key| encrypted_attributes[attribute].delete(key) }
+                self.class.decrypt(attribute, public_send("#{attribute_options[:attribute]}_was"), evaluated_options)
+              end
             end
+          end
+
+          def reflect_on_aggregation_with_attr_encrypted(attribute)
+            encrypted_reflections[attribute.to_sym] || reflect_on_aggregation_without_attr_encrypted(attribute)
+          end
 
-            alias_method "#{attr}_before_type_cast", attr
```

Comment:
Is it necessary to remove this? I think it makes sense to do the decryption before typecasting.